fix for GetAccountsAsync() w/b2clogin.com authorities

jennyf19 opened the pull request:
- move b2c instance discovery to new method  to access in token cache
- in token cache, do instance discovery w/b2c authority.
- refactor b2c ui tests -> doing one test per provider, which includes AT interactive and silent
- add ui tests for FB provider
- add IdentityProvider to userQP's to differentiate between the different idps

Tested manually and w/the new UI tests.

jennyf19 committed:
fix for GetAccountsAsync() w/b2clogin.com authorities
- move b2c instance discovery to new method  to access in token cache
- in token cache, do instance discovery w/b2c authority.
- refactor b2c ui tests -> doing one test per provider, which includes AT interactive and silent
- add ui tests for FB provider
- add IdentityProvider to userQP's to differentiate between the different idps

jennyf19 commented on msal/src/Microsoft.Identity.Client/TokenCache.cs:
```diff
@@ -694,6 +694,14 @@ internal MsalAccountCacheItem GetAccountCacheItem(MsalAccountCacheKey msalAccoun
+                    HttpManager,
+                    TelemetryManager,
+                    requestContext).ConfigureAwait(false);
+            }
```
The list of environment alias will contain the authority.Host (ex. msidlabb2clogin.com) and the aliases from the instance discovery.  #Closed

jennyf19 commented on msal/tests/Test.MSAL.NET.UIAutomation/XamarinDroidMSALTests.cs:
```diff
@@ -104,42 +104,60 @@ public void PromptBehavior_Consent_SelectAccount()
-        /// B2C aquire token flow with local account and b2clogin.com authority
+        /// B2C acquire token with Facebook provider
+        /// b2clogin.com authority
+        /// with subsequent silent call
```
consolidating tests...since the silent test has to do an interactive call to get the token...two tests in one (one interactive and one silent). #Closed

jennyf19 committed:
fix integration test failure

jennyf19 committed:
rename to match lab api

jennyf19 committed:
Add Google UI test for b2clogin.com and login.microsoftonline.com

henrik-me commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -68,16 +68,28 @@ internal B2CAuthority(string authority, bool validateAuthority)
             }
 
-            var metadata = await AadInstanceDiscovery
+            InstanceDiscoveryMetadataEntry metadata = await GetB2CInstanceDiscoveryMetadataEntryAsync(
```
>GetB2CInstanceDiscoveryMetadataEntryAsync [](start = 60, length = 41)

Thinking the instance discovery point should have a virtual method or alternatively that the instance discovery object should be injected....

Also not sur what actual change is made here. #Resolved

bgavrilMS replied:
Currently there is only one AADInstanceDiscovery. This is a singleton that maintains a cache. The B2CAuthority is a type of AADAuthority so it somewhat seems ok to be able to call AADInstanceDiscovery. #Resolved

jennyf19 replied:
outdated due to latest changes.

---
In reply to: [233710236](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r233710236) [](ancestors = 233710236)

jennyf19 replied:
outdated due to latest changes.

---
In reply to: [234182155](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234182155) [](ancestors = 234182155)

henrik-me commented on core/tests/Test.Microsoft.Identity.Core.UIAutomation/UserInformationFieldIds.cs:
```diff
@@ -51,11 +51,21 @@ public void DetermineFieldIds(LabUser user)
                         break;
                 }
             }
-            else if (user.UserType == UserType.B2C)
```
>(user.UserType == UserType.B2C) [](start = 20, length = 31)

assuming user.UserType is still B2C?  #Closed

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/B2CProvider.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Test.Microsoft.Identity.LabInfrastructure
+{
+    public enum B2CProvider
```
B2CIdentityProvider ? #Resolved

jennyf19 replied:
i was matching what the lab returned, but asked them to change to this. because it's clearer.

---
In reply to: [233710705](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r233710705) [](ancestors = 233710705)

henrik-me commented on msal/src/Microsoft.Identity.Client/TokenCache.cs:
```diff
@@ -685,7 +685,7 @@ internal MsalAccountCacheItem GetAccountCacheItem(MsalAccountCacheKey msalAccoun
-            if (authorityType == Core.Instance.AuthorityType.Aad || authorityType == Core.Instance.AuthorityType.B2C)
+            if (authorityType == Core.Instance.AuthorityType.Aad)
             {
                 instanceDiscoveryMetadata = await AadInstanceDiscovery.Instance.GetMetadataEntryAsync(
```
>AadInstanceDiscovery.Instance [](start = 50, length = 29)

this is actually the type imo, which should be set to a B2C Instance discovery. It seems odd to have to if here... the correct type should be injected #Closed

bgavrilMS replied:
Yes, I agree, it is a code smell to have an inheritance hierarchy and to have to call GetType() on it. I don't think we need a B2CInstanceDiscovery though, because it seems to me to that the only difference between B2CInstanceDiscovery and AADInstanceDiscovery is the authority uri. How about

Solution 1

Move all the logic to the helper class AADInstanceDiscovery. It can inspect the authority type and decide how to do the discovery.

Solution 2
- Expose a method GetCachedOrDiscoverAuthorityMetaDataAsync on the Authority class
- Each authority type knows how to discover its own metadata, using the helper class AADInstanceDiscovery
- Authorities other than AAD and B2C do nothing  #Resolved

henrik-me commented:
great to see B2C e2e tests coming alive!  #Closed

henrik-me commented:
IMO, you are running into issues here because the Authority class has too many responsibilities. Refactoring this will likely require a few changes to the class inheritance and injected objects to ensure a better separation of concern. #Resolved

henrik-me reviewed (changes requested):
:clock1:

jennyf19 committed:
Address some PR comments

bgavrilMS commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/B2CIdentityProvider.cs:
```diff
@@ -27,10 +27,11 @@
-        Local = 0,
-        Facebook = 1,
-        Google = 2
+        None = 0,
```
Please add comments to disambiguate None and Local #Resolved

bgavrilMS commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -68,16 +68,28 @@ internal B2CAuthority(string authority, bool validateAuthority)
+            ITelemetryManager telemetryManager,
+            RequestContext requestContext)
+        {
+             InstanceDiscoveryMetadataEntry metadata = await AadInstanceDiscovery
```
Inheritance chains are difficult to maintain, because you have to keep track of what each base class is doing, if you are calling into the base class itself etc. I much prefer interfaces and encapsulation...

In this specific case, I see 2 problems: 

- InTrustedHostList is a private instance member in this class, and static in base class. You are actually calling into the base class, but this is difficult to comprehend. I would rename the static method as IsTrustedAADHost(), or move it to a separate class. 
- There is confusion in having a CannonicalAuthority field in the inheritance chain, which is updated A LOT during the lifetime of the class. And you add a GetCanonicalAuthorityB2C() which seems to be doing more changes. Can't all of these be consolidated ? #Resolved

jennyf19 replied:
outdated due to latest changes. (but i appreciate and did read/think about your comment :) )

---
In reply to: [234184038](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234184038) [](ancestors = 234184038)

jennyf19 committed:
remove instance discover for b2c
set validate authority to false for b2c
fix up ui tests for b2c
removed unnecessary utests for b2c

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/b2cLoginGetAccounts

jennyf19 committed:
switching to lab beta slice for changes made to b2cProvider

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/LabServiceApi.cs:
```diff
@@ -87,7 +97,7 @@ private LabResponse GetLabResponseFromAPI(UserQueryParameters query)
             }
 
-            LabResponse response = JsonConvert.DeserializeObject<LabResponse>(result);
+LabResponse response = JsonConvert.DeserializeObject<LabResponse>(result);
```
nit: indentation #Closed

jennyf19 replied:
done.

---
In reply to: [234387018](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234387018) [](ancestors = 234387018)

henrik-me commented on core/tests/Test.Microsoft.Identity.LabInfrastructure/LabServiceApi.cs:
```diff
@@ -71,12 +71,22 @@ private LabResponse GetLabResponseFromAPI(UserQueryParameters query)
+                queryDict.Add("b2cProvider", "google");
+            }
+
+            UriBuilder uriBuilder = new UriBuilder("http://api.msidlab.com/api/userbeta");
```
>UriBuilder uriBuilder = new UriBuilder("http://api.msidlab.com/api/userbeta"); [](start = 12, length = 78)

will the beta endpoint continue to be available? an ETA on this going to the non beta end-point? #Closed

jennyf19 replied:
i thought it would be today, so hopefully monday. I wanted to run a release build to test some b2c samples. this is temporary.

---
In reply to: [234387151](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234387151) [](ancestors = 234387151)

jennyf19 committed:
indentation

henrik-me commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -99,11 +103,7 @@ private bool InTrustedHostList(string host)
-            // To support host we would need to make the call to the tenant discovery end point, 
-            // which is not supported for b2clogin.com
-            string authorityUri = string.Format(CultureInfo.InvariantCulture, MicrosoftOnline + b2cAuthority.AbsolutePath + OpenIdConfigurationEndpoint);
+            string authorityUri = string.Format(CultureInfo.InvariantCulture, new Uri(CanonicalAuthority) + OpenIdConfigurationEndpoint);
```
>ew Uri(CanonicalAuthority) [](start = 79, length = 26)

IMO, you should be more specific on which value you are using from the uri object instead of relying on .ToString() #Closed

jennyf19 replied:
not sure if this is what you mean. 

---
In reply to: [234387424](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234387424) [](ancestors = 234387424)

henrik-me replied:
great, you did what I was asking for :)

---
In reply to: [234388886](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388886) [](ancestors = 234388886,234387424)

jennyf19 replied:
#ItsAMiracle

---
In reply to: [234388977](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388977) [](ancestors = 234388977,234388886,234387424)

henrik-me commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -99,11 +103,7 @@ private bool InTrustedHostList(string host)
-            // To support host we would need to make the call to the tenant discovery end point, 
-            // which is not supported for b2clogin.com
-            string authorityUri = string.Format(CultureInfo.InvariantCulture, MicrosoftOnline + b2cAuthority.AbsolutePath + OpenIdConfigurationEndpoint);
+            string authorityUri = string.Format(CultureInfo.InvariantCulture, new Uri(CanonicalAuthority) + OpenIdConfigurationEndpoint);
```
>authorityUri [](start = 19, length = 12)

nit: not sure it is worth it as this may only be used once.... it may be worth caching for this instance of the class.  #Closed

jennyf19 committed:
add comments

henrik-me commented on msal/src/Microsoft.Identity.Client/TokenCache.cs:
```diff
@@ -684,15 +688,22 @@ internal MsalAccountCacheItem GetAccountCacheItem(MsalAccountCacheKey msalAccoun
+            {
+                return instanceDiscoveryMetadata;
             }
             return instanceDiscoveryMetadata;
```
why not just return null for these 2 cases. only for AAD will it be set to anything besides null #Closed

jennyf19 replied:
when i tested w/login.microsoftonline.com, the cache key was login.windows.net (or something like that), so we need to do the instance discovery for the login.microsoftonline.com authorities, but not the b2clogin.com ones

---
In reply to: [234387980](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234387980) [](ancestors = 234387980)

henrik-me replied:
not what I meant. You still  have the instance discovery happening for the AAD case above. You are actively not doing the instance discovery here. This place the variable will always be null.

---
In reply to: [234388128](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388128) [](ancestors = 234388128,234387980)

jennyf19 replied:
oh i get it.

---
In reply to: [234388879](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388879) [](ancestors = 234388879,234388128,234387980)

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:
```diff
@@ -219,5 +287,43 @@ public void PerformB2CLocalAccountSignInFlow(ITestController controller, LabUser
+            controller.EnterText(userInformationFieldIds.PasswordInputId, LabUserHelper.GetUserPassword(user), XamarinSelector.ByHtmlIdAttribute);
+
+            controller.Tap(userInformationFieldIds.SignInButtonId, XamarinSelector.ByHtmlIdAttribute);
+        }
```
There are small differences between these 3 flows. Consider refactoring allowing for a section of specifc provider code and then some template/resuable before and after code.. #Closed

jennyf19 replied:
done. was easier to fix up once i had all the tests done. thanks.

---
In reply to: [234388341](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388341) [](ancestors = 234388341)

henrik-me commented on msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/InstanceTests/B2cAuthorityTests.cs:
```diff
@@ -76,34 +76,7 @@ public void NotEnoughPathSegmentsTest()
-
-        [TestMethod]
-        [TestCategory("B2CAuthorityTests")]
-        public void ValidationEnabledNotSupportedTest()
```
>public void ValidationEnabledNotSupportedTest() [](start = 8, length = 47)

you should probably keep a test saying even if you set it to true, no authority validation is happening? #Closed

jennyf19 replied:
added 2 tests, one for b2clogin.com and one for login.microsoftonline.com w/the json response for openid w/validate=true.

---
In reply to: [234388511](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388511) [](ancestors = 234388511)

henrik-me commented on msal/tests/Test.MSAL.NET.Unit.net45/PublicApiTests/PublicClientApplicationTests.cs:
```diff
@@ -1300,71 +1297,6 @@ public void B2CAcquireTokenTest()
-                catch (Exception exc)
-                {
-                    Assert.AreEqual(exc.InnerException.Message, CoreErrorMessages.UnsupportedAuthorityValidation);
-                }
```
this should now work and not throw an exception, right? Why not keep the test? same below #Resolved

jennyf19 replied:
yep. i put them back. 

---
In reply to: [234388585](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234388585) [](ancestors = 234388585)

jennyf19 committed:
pr comments

jennyf19 committed:
remove extra check in token cache
refactor b2c ui tests
add b2c utests w/authority validation false
add openid config for b2clogin

jennyf19 commented:
i agree, but think this is irrelevant w/latest changes. something to add to our todo list?

---
In reply to: [438916765](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#issuecomment-438916765) [](ancestors = 438916765)

jennyf19 committed:
adding back utests for b2c validate=true

henrik-me commented on msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:
```diff
@@ -197,27 +141,176 @@ private void ValidateUiBehaviorString(string uiBehavior)
             }
         }
 
+        #region B2C UI Automation
```
>#region B2C UI Automation [](start = 8, length = 25)

nit: I'm personally not a fan of #regions... however happy to have those in case others are in favor #Resolved

henrik-me commented:
>         public bool isB2CloginAuthority;

imo, this var should be passed around and should not be a class field (aside if being a field it should be _isXYZ #Resolved

---
Refers to: msal/tests/Test.MSAL.NET.UIAutomation/MSALMobileTestHelper.cs:45 in 21688c5. [](commit_id = 21688c5323b9d17083c4d7b4ffb6897cb191db92, deletion_comment = False)

henrik-me commented on msal/tests/Test.MSAL.NET.Unit.net45/CoreTests/InstanceTests/B2cAuthorityTests.cs:
```diff
@@ -79,28 +82,80 @@ public void NotEnoughPathSegmentsTest()
+                    }).GetAwaiter().GetResult();
+
+                Assert.AreEqual(
+                    "https://mytenant.com.b2clogin.com/6babcaad-604b-40ac-a9d7-9fd97c0b779f/oauth2/v2.0/authorize?p=my-policy",
```
>my-policy [](start = 116, length = 9)

We should work with the function where the policy is in the url after the tenant and not added as a parameter. This is not oidc compliant. I'm actually puzzled why you get the parameterized response back. While this will work for AAD/B2C, I believe we should look into changing this. IMO, this work can be done as a new task. #Resolved

henrik-me commented on msal/tests/Test.MSAL.NET.Unit.net45/Resources/OpenidConfiguration-B2CLogin.json:
```diff
@@ -0,0 +1,17 @@
+﻿{
+  "authorization_endpoint": "https://mytenant.com.b2clogin.com/6babcaad-604b-40ac-a9d7-9fd97c0b779f/oauth2/v2.0/authorize?p=my-policy",
+  "token_endpoint": "https://mytenant.com.b2clogin.com/6babcaad-604b-40ac-a9d7-9fd97c0b779f/oauth2/v2.0/token?p=my-policy",
```
>my-policy [](start = 112, length = 9)

ah, this is likely what needs to be updated. the policy should be part of the url right after the tenant afaik #Resolved

jennyf19 replied:
i fixed this, but one of the conf client tests is checking for the policy as a qp, so I made another json file for that case. I didn't want to change the unit test at this point. also, i noticed these files have to be in the adal tests too in order for the utests to pass...which seems odd. I'll sync w/bogdan on that.

---
In reply to: [234397668](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234397668) [](ancestors = 234397668)

jennyf19 committed:
address PR comments
- remove qp from b2c config
- add conf client b2c config w/qp for utest
- remove class variable and pass around isB2CLoginAuthority instead

henrik-me commented on core/src/Instance/B2CAuthority.cs:
```diff
@@ -46,6 +46,7 @@ internal B2CAuthority(string authority, bool validateAuthority)
             : base(authority, validateAuthority)
         {
             AuthorityType = AuthorityType.B2C;
+            validateAuthority = false;
```
>validateAuthority = false; [](start = 12, length = 26)

consider adding a comment as to why we decided to set this to false. #Resolved

jennyf19 committed:
add comment

henrik-me reviewed (approved):
:shipit:
